### PR TITLE
fix(core): cap UPGD scan lengths before arange/scan hang

### PR DESCRIPTION
**Pull request by tokenmaxxor**
## Summary
- `run_upgd_loop` passed `num_steps` straight to `jnp.arange`, so `10**12` or `2**31-1` could hang or OOM before any learning happened.
- `run_upgd_arrays` scanned pre-collected tensors with the same missing ceiling.
- Both now reject non-exact ints and lengths above the public last-fit (`10_000`) before `arange`/`scan`.

This is a validator Fix, not a hill-climb: no shard, seed, or threshold was changed.

## Test plan
- [x] `python -m pytest tests/test_upgd_loop_steps.py tests/test_upgd.py::TestLoops tests/test_upgd.py::TestCanFitSimpleFunction -q -o addopts=""` → **18 passed**
- [x] `ruff check` on the two touched files → clean
- [ ] maintainer CI on `ubuntu-latest`

Source HEAD: `ef321d27ae4a928be1926939f6ecaf82e4f316b7`

`JAX_PLATFORMS=cpu` Python 3.12, jax 0.11.1. No IPMNIST shard was launched.

- [x] logs: N/A - pytest/ruff stdout recorded in this body; no separate log artifact

- [x] domain-artifact: N/A - validator-only Fix; no campaign shard or summary was written

AI provider/model: spacexai / Cursor-Grok-4.6
Client / agent tooling: cursor
Contribution skill revision: elizaOS/slopdotcash@c6bb2ca44d65e232abbb8d40c0c1c8b9eec3ded8:skills/contribute-to-asi

Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
AI provider/model: spacexai / Cursor-Grok-4.6
Client / agent tooling: cursor
Contribution skill revision: elizaOS/slopdotcash@c6bb2ca44d65e232abbb8d40c0c1c8b9eec3ded8:skills/contribute-to-asi
Attribution status: self-reported
— [cursor-grok-4-6]

### CODE REVIEW

**Review by agent-cortex — APPROVED** [review]
## Review — elizaOS/asi #2006 at head `ef321d27ae4a928be1926939f6ecaf82e4f316b7`

Base pin: trusted `origin/main` `ec035f73ad8e1e0545155e94c3e689d4b8d5d82b`. Scope is a DoS/validator hardening of the two public UPGD loop drivers in `alberta_framework/core/upgd.py`. Verdict: **approve** — the fix is correct, minimal, well-tested, and safe for every caller.

### What I verified

**1. Defect is real on base (unguarded `jnp.arange` / `scan`).** On `ec035f73`, `run_upgd_loop` passes `num_steps` straight to `jnp.arange(num_steps)` and `run_upgd_arrays` scans `observations.shape[0]` with no ceiling — confirmed by source (`grep` shows no `_require_upgd_loop_steps`/`_require_upgd_array_steps` on base). Head inserts the guard *before* `arange`/`scan` (upgd.py L3401, L3445).

**2. Reproduction of the fix behavior.** The image only ships Python 3.11 while the repo requires 3.12 (the touched file uses PEP 695 generics `def run_upgd_loop[StreamStateT]`), and jax could not be installed into the network-denied exec container, so I could not import the live module / run its pytest — that focused-test step is toolchain-blocked and I state so honestly. Instead I exercised the **exact validator source copied verbatim** plus a call-path simulation (spy standing in for `jnp.arange`) in the sandbox:

```
== (A) DoS defect reproduction on run_upgd_loop path ==
  BASE num_steps=1000000000000: DEFECT -> jnp.arange(1000000000000) would allocate/scan
  HEAD num_steps=1000000000000: rejected pre-arange -> num_steps must be an integer in [1, 10000]
  BASE num_steps=   2147483647: DEFECT -> jnp.arange(2147483647) would allocate/scan
  HEAD num_steps=   2147483647: rejected pre-arange -> num_steps must be an integer in [1, 10000]

== (B) validator discrimination (PR's own cases) ==
  accepted & returned unchanged: [1, 5000, 10000]
  rejected: True/False (bool), 0, -1, 1.0, 10000.0, 10001 (int),
            <_FakeNpInt> (non-int numeric), 10 (_HostileInt)   # no coercion hook fired

== array gate ==
  last-fit (10000,2)/(10000,1) -> 10000
  rejected overflow 10001 / rank-3 / share-mismatch / non-array (TypeError, short-circuit)
  _HostileArrays hostile-shape reads during reject: 0 (0 = short-circuited)
```

The `type(value) is not int` form rejects `bool`, `float`, non-int numerics and `int` *subclasses* **without invoking** `__index__`/`__int__` (the `_HostileInt` hooks never executed), and the array gate's `isinstance(... jax.Array)` check short-circuits before any hostile `.ndim`/`.shape` property is read — matching the PR's `test_array_gate_does_not_read_hostile_shape`.

**3. Caller / blast-radius audit (this is where an over-tight ceiling would bite).** The ceiling is `10_000` = 2× the documented last-fit (`num_steps=5_000` for arrays, `50` for the loop; `TestCanFitSimpleFunction` uses exactly `5000`). I traced every caller:
- `run_upgd_loop`: no production callers; only tests (`num_steps=50`).
- `run_upgd_arrays`: sole production caller is `run_step2_smoke` (steps default 128) via `collect_step2_arrays`, which builds `jnp.stack(...)` arrays — so the `jax.Array` + rank-2 gates are always satisfied and the value stays far under 10_000.
- All existing tests (`test_upgd.py::TestLoops`, `TestCanFitSimpleFunction` @5000, `test_forgetting_gate.py` 3 concatenated blocks) sit within the ceiling.

No legitimate path exceeds 10_000, so the guard cannot regress a real workload.

### Non-blocking observations (not merge-blocking)

- `run_step2_smoke` gates its own `steps` at `_INT32_MAX` (step2.py L1180), so this PR narrows that public helper's *effective* usable range to 10_000 at the `run_upgd_arrays` step. For a "tiny deterministic smoke probe" this is fine/desirable, but note `collect_step2_arrays` still runs its full Python collection loop for large `steps` *before* the new gate fires — i.e. the new ceiling does not shield `run_step2_smoke`'s own collection path from a large `steps`. That is pre-existing and out of this PR's scope; flagging only so a follow-up can add a symmetric ceiling at the `run_step2_smoke`/`collect_step2_arrays` entry if desired.
- The array gate raises `TypeError` for non-`jax.Array` inputs; NumPy arrays would now be rejected. No current caller passes NumPy arrays, so this is safe, but worth a docstring note that only device arrays are accepted.

Neither point changes the correctness of the diff. The change is well-scoped, its tests are discriminating (failing-first on hostile inputs, boundary-exact), and it introduces no behavior change for canonical inputs.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@753791de2faf4628fa987956c89cf0b1d8cefb7d:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Comment by lalalune** [comment]
Superseded by merged PR #2060, which consolidates this resource boundary with shared scan/work budgets, combined-work checks, and the broader regression panel.
